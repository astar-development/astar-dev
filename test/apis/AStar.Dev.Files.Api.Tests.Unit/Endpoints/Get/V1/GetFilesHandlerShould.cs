using AStar.Dev.Test.Helpers.Minimal.Api;
using AStar.Dev.Utilities;
using DbContextHelpers.Fixtures;
using Microsoft.AspNetCore.Http;
using NSubstitute;

namespace AStar.Dev.Files.Api.Endpoints.Get.V1;

public class GetFilesHandlerShould : IClassFixture<FilesContextFixture>
{
    private readonly FilesContextFixture mockFilesContextFactory;
    private readonly TimeProvider        mockTimeProvider;

    public GetFilesHandlerShould(FilesContextFixture mockFilesContextFactory)
    {
        this.mockFilesContextFactory = mockFilesContextFactory;
        mockTimeProvider             = Substitute.For<TimeProvider>();
        mockTimeProvider.GetUtcNow().Returns(new DateTimeOffset(new (2025, 7, 13, 1, 2, 3, DateTimeKind.Utc)));
    }

    [Fact]
    public async Task ReturnBadRequestWhenDuplicatesAreRequested()
    {
        await using var mockContext = mockFilesContextFactory.Sut;
        var             sut         = new GetFilesHandler();

        var response = await sut.HandleAsync(new(), mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ShouldBe(Results.BadRequest());
    }

    [Theory]
    [InlineData(@"c:\temp", true,  0, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                              SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                  expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }

    [Theory]
    [InlineData(@"c:\temp", false,  0, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters2(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                               SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                   expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }

    [Theory]
    [InlineData(@"c:\temp", true,  5, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters3(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                               SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                   expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }

    [Theory]
    [InlineData(@"c:\temp", false, 5, true,  "",  SortOrder.NameAscending,  SearchType.All, 1, 10, 10)]
    public async Task ReturnTheExpectedStructureWhenCalledWithValidParameters4(string    directoryName, bool recursive, int excludeViewedWithinDays, bool includeMarkedForDeletion, string searchText,
                                                                               SortOrder sortOrder,     SearchType searchType, int currentPage, int itemsPerPage, int
                                                                                   expectedCount)
    {
        await using var mockContext = mockFilesContextFactory.SutWithFileDetails;
        var             sut         = new GetFilesHandler();

        var response =
            await
                sut.HandleAsync(new()
                                {
                                    SearchType               = searchType,
                                    SearchFolder             = directoryName,
                                    Recursive                = recursive,
                                    ExcludeViewedWithinDays  = excludeViewedWithinDays,
                                    IncludeMarkedForDeletion = includeMarkedForDeletion,
                                    SortOrder                = sortOrder,
                                    SearchText               = searchText,
                                    CurrentPage              = currentPage,
                                    ItemsPerPage             = itemsPerPage
                                }, mockContext, mockTimeProvider, "Test User", CancellationToken.None);

        response.ResultStatusCode().ShouldBe(200);
        var getFilesResponses = response.ResultValue<IList<GetFilesResponse>>()!;
        getFilesResponses.Count.ShouldBe(expectedCount);
        getFilesResponses.ToJson().ShouldMatchApproved();
    }
}
