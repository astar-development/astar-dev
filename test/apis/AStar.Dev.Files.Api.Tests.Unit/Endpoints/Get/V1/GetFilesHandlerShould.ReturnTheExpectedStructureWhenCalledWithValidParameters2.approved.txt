[
  {
    "directoryName": "c:\\temp",
    "fileName": "CodeMaid.config",
    "fileHandle": "c:\\temp-CodeMaid.config-1"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "FileAccessDetails.csv",
    "fileHandle": "c:\\temp-FileAccessDetails.csv-3"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "Files.csv",
    "fileHandle": "c:\\temp-Files.csv-4"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "Files.db",
    "fileHandle": "c:\\temp-Files.db-5"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "ef-core-6-encapsulating-usage.zip",
    "fileHandle": "c:\\temp-ef-core-6-encapsulating-usage.zip-2"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "wallhaven-8568w2.png",
    "fileHandle": "c:\\temp-wallhaven-8568w2.png-6"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "wallhaven-85819j.jpg",
    "fileHandle": "c:\\temp-wallhaven-85819j.jpg-7"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "wallhaven-eo3198.jpg",
    "fileHandle": "c:\\temp-wallhaven-eo3198.jpg-8"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "wallhaven-ex221l.jpg",
    "fileHandle": "c:\\temp-wallhaven-ex221l.jpg-9"
  },
  {
    "directoryName": "c:\\temp",
    "fileName": "wallhaven-expvqk.jpg",
    "fileHandle": "c:\\temp-wallhaven-expvqk.jpg-10"
  }
]